File: word-count/WordCount.hs
module WordCount where

import Data.Char (isAlphaNum, toLower)
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map

wordCount :: String -> Map String Int
wordCount = foldr (f . filter isAlphaNum) Map.empty . words'
  where f []     = id
        f x      = Map.insertWith (+) (map toLower x) 1
        words' s = case dropWhile p s of
                        "" -> []
                        s' -> w : words' s'' where (w, s'') = break p s'
          where p = not . isAlphaNum

File: word-count/word_count.erl
-module(word_count).

-export([count/1]).

count(String) ->
  %% So lame that I have to remove_empty. Come on, re:split/3..
  %% http://erlang.org/pipermail/erlang-questions/2013-December/076202.html
  Words = remove_empty(re:split(String, "[^0-9A-Za-z]+",[{return,list},trim])),
  F     = fun(W, D) -> dict:update_counter(string:to_lower(W), 1, D) end,
  lists:foldl(F, dict:new(), Words).

remove_empty(List) ->
  Pred = fun(X) ->
             case X of
               [] -> false;
               _  -> true
             end
         end,
  lists:filter(Pred, List).
